src/twisted/protocols/_smb/core.py

```diff
@@ -0,0 +1,760 @@
+    credit_charge = short()
+    status = medium()
+    command = short()
+    credit_request = short()
+    flags = medium()
+    next_command = medium()
+    message_id = long()
+    reserved = medium()
+    tree_id = medium()
+    session_id = long()
+    signature = octets(16)
+    async_id: int = attr.ib(default=0)
+
+
+@attr.s
+class HeaderAsync:
+    magic = octets(default=SMB2_MAGIC)
+    size = short()
+    credit_charge = short()
+    status = medium()
+    command = short()
+    credit_request = short()
+    flags = medium()
+    next_command = medium()
+    message_id = long()
+    async_id = long()
+    session_id = long()
+    signature = octets(16)
+    tree_id: int = attr.ib(default=0)
+
+
+def packetReceived(packet: _base.SMBPacket) -> None:
+    """
+    receive a SMB packet with header. Unpacks the
+    header then calls the appropriate smb_XXX
+    method with data beyond the header.
+
+    @param packet: the raw packet
+    @type packet: L{_base.SMBPacket}
+    """
+    offset = 0
+    isRelated = True
+    while isRelated:
+        protocol_id = packet.data[offset : offset + len(SMB2_MAGIC)]
+        if protocol_id == SMB1_MAGIC:
+            # its a SMB1 packet which we dont support with the exception
+            # of the first packet, we try to offer upgrade to SMB2
+            if packet.ctx.get("avatar") is None:
+                log.debug("responding to SMB1 packet")
+                negotiateResponse(packet)
+            else:
+                packet.close()
+                log.error("Got SMB1 packet while logged in")
+            return
+        elif protocol_id != SMB2_MAGIC:
+            packet.close()
+            log.error("Unknown packet type")
+            log.debug("packet data {data!r}", data=packet.data[offset : offset + 64])
+            return
+        packet.hdr, o2 = _base.unpack(HeaderSync, packet.data, offset, _base.OFFSET)
+        isAsync = (packet.hdr.flags & FLAG_ASYNC) > 0
+        isRelated = (packet.hdr.flags & FLAG_RELATED) > 0
+        isSigned = (packet.hdr.flags & FLAG_SIGNED) > 0
+        # FIXME other flags 3.1 or too obscure
+        if isAsync:
+            packet.hdr = _base.unpack(HeaderAsync, packet.data, offset)
+        if isRelated:
+            this_packet = packet.data[offset : offset + packet.hdr.next_command]
+        else:
+            this_packet = packet.data[offset:]
+        flags_desc = ""
+        if isAsync:
+            flags_desc += " ASYNC"
+        if isRelated:
+            flags_desc += " RELATED"
+        if isSigned:
+            flags_desc += " SIGNED"
+        log.debug(
+            """
+HEADER
+------
+protocol ID     {pid!r}
+size            {hs}
+credit charge   {cc}
+status          {status}
+command         {cmd!r} {cmdn:02x}
+credit request  {cr}
+flags           0x{flags:04x}{flags_desc}
+next command    0x{nc:x}
+message ID      0x{mid:x}
+session ID      0x{sid:x}
+async ID        0x{aid:x}
+tree ID         0x{tid:x}
+signature       {sig}""",
```

Comment:
FYI, you can avoid this type of ugly outdent by using `textwrap.dedent` on an indented string.